corehq/apps/locations/util.py

```diff
@@ -179,55 +199,68 @@ def get_consumption(self, loc):
-
-            uncategorized_keys.update(uncategorized_data)
-
-            loc_dict = {
-                'location_id': loc.location_id,
-                'site_code': loc.site_code,
-                'name': loc.name,
-                'parent_site_code': loc.parent.site_code if loc.parent else '',
-                'latitude': loc.latitude or '',
-                'longitude': loc.longitude or '',
-                'data': model_data,
-                'uncategorized_data': uncategorized_data,
-                'consumption': self.get_consumption(loc),
-                LOCATION_SHEET_HEADERS['external_id']: loc.external_id,
-                LOCATION_SHEET_HEADERS['do_delete']: ''
-            }
-            tab_rows.append(dict(flatten_json(loc_dict)))
-
-        header_keys = ['location_id', 'site_code', 'name', 'parent_code',
-                       'latitude', 'longitude', 'external_id', 'do_delete']
-        tab_headers = [LOCATION_SHEET_HEADERS[h] for h in header_keys]
-
-        def _extend_headers(prefix, headers):
-            tab_headers.extend(json_to_headers(
-                {prefix: {header: None for header in headers}}
-            ))
-        _extend_headers('data', (f.slug for f in self.data_model.fields))
-        _extend_headers('uncategorized_data', uncategorized_keys)
-        if self.include_consumption_flag and loc_type.name not in self.administrative_types:
-            _extend_headers('consumption', self.product_codes)
-
-        sheet_title = loc_type.code
-
-        return (sheet_title, {
-            'headers': tab_headers,
-            'rows': tab_rows,
-        })
+    def get_headers(self):
+        headers = OrderedDict([
+            ('types', [list(LOCATION_TYPE_SHEET_HEADERS.values())])
+        ])
+        for loc_type in self.location_types:
+            additional_headers = []
+            additional_headers.extend(self._prefix_headers('data', (f.slug for f in self.data_model.fields)))
+            if self.include_consumption_flag and loc_type.name not in self.administrative_types:
+                additional_headers.extend(self._prefix_headers('consumption', self.product_codes))
+            additional_headers.append(LOCATION_SHEET_HEADERS_OPTIONAL['uncategorized_data'])
+            additional_headers.append(LOCATION_SHEET_HEADERS_OPTIONAL['delete_uncategorized_data'])
+
+            headers[loc_type.code] = [LOCATION_SHEET_HEADERS_BASE.values() + additional_headers]
+
+        return list(headers.items())
+
+    def write_data(self, writer):
+        self._write_type_sheet(writer)
+        for location_type in self.location_types:
+            self._write_locations(writer, location_type)
+
+    def _write_locations(self, writer, location_type):
+        include_consumption = self.include_consumption_flag and location_type.name not in self.administrative_types
+        query = SQLLocation.active_objects.filter(
+            domain=self.domain,
+            location_type__name=location_type.name
```

Comment:
I think this should be `location_type=location_type` or `location_type_id=location_type.pk` so it doesn't have to do a join.

Reply:
407632c